src/reflect/scala/reflect/internal/tpe/TypeMaps.scala

```diff
@@ -698,12 +697,21 @@ private[internal] trait TypeMaps {
         tp
     }
 
-    @tailrec private def subst(tp: Type, sym: Symbol, from: List[Symbol], to: List[T]): Type = (
-      if (from.isEmpty) tp
-      // else if (to.isEmpty) error("Unexpected substitution on '%s': from = %s but to == Nil".format(tp, from))
-      else if (matches(from.head, sym)) toType(tp, to.head)
-      else subst(tp, sym, from.tail, to.tail)
-      )
+    private[this] def subst(tp: Type, sym: Symbol): Type = {
+      val found = find(sym)
+      if (found == null) tp else toType(tp, found)
+    }
+
+    protected def find(sym: Symbol): T
+
+    protected def findIndex(sym: Symbol): Int = {
+      @tailrec def loop(syms: List[Symbol], ix: Int, sym: Symbol): Int = syms match {
+        case Nil => -1
+        case x :: _ if matches(x, sym) => ix
+        case _ :: xs => loop(xs, ix+1, sym)
+      }
+      loop(from, 0, sym)
+    }
```

Comment:
We could also add a fast path to return `-1` by checking `fromIdMax`, `fromIdMin`, `fromSize` as we do in `fromContains` below.

Reply:
I am not sure that is completely correct... the thing is, the search is not directly based on a concrete comparison, but on a non-final method `matches`. This method has a default implementation which is reference equality, but some instances of the class override it to `deSkolemize` the symbols.